Add star rating to the beatmap metadata in player loading screen

frenzibyte opened the pull request:
Closes #12414 

- [x] Depends on #12747

Went with the simplest way of adding a `StarRatingDisplay`, as desired in the issue thread, looks harmonious enough.

https://user-images.githubusercontent.com/22781491/117533089-72a10100-aff3-11eb-9fc2-664cd60497a0.mp4

(in case it isn't obvious, this is from a test scene, not from game)

frenzibyte committed:
Add star rating display underneath the beatmap metadata

frenzibyte committed:
Allow null logo facade

frenzibyte committed:
Add visual test scene

frenzibyte committed:
Pass ruleset info to constructor instead

Follows the way working beatmap is passed, not sure why mods are passed as a bindable though, don't wanna bother too much with that.

frenzibyte committed:
Refactor `StarRatingDisplay` to be mutable with a current bindable

frenzibyte committed:
Use `BeatmapDifficultyCache.GetBindableDifficulty(...)` instead

frenzibyte committed:
Remove no longer necessary ruleset info requirement

frenzibyte committed:
Fix test not handling 0 beatmap sets

bdach commented on osu.Game.Tests/Visual/Ranking/TestSceneStarRatingDisplay.cs:
```diff
@@ -25,8 +30,14 @@ public TestSceneStarRatingDisplay()
             };
-        }
+
+            Scheduler.AddDelayed(() =>
+            {
+                changingStarRating.Current.Value = new StarDifficulty(RNG.NextDouble(0, 10), RNG.Next());
+            }, 500, true);
+        });
```
not sure why this is a delayed schedule inside a schedule, while this test scene could just use normal test steps

frenzibyte replied:
sure.

bdach commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapMetadataDisplay.cs:
```diff
@@ -0,0 +1,81 @@
+
+namespace osu.Game.Tests.Visual.SongSelect
+{
+    [System.ComponentModel.Description("player loader beatmap metadata")]
```
i don't think this description is very useful

frenzibyte replied:
I've used this because `BeatmapMetadataDisplay` sounded a bit off, but I guess this test scene should be considered irrelevant from any mentioning of player loaders, will remove.

bdach replied:
that, and the name of the subject under test is in the test class name. there is no need to repeat information.

bdach commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapMetadataDisplay.cs:
```diff
@@ -0,0 +1,81 @@
+        [Resolved]
+        private BeatmapManager manager { get; set; }
+
+        private IReadOnlyList<Mod> randomMods => Ruleset.Value.CreateInstance()
+                                                        .GetAllMods()
+                                                        .OrderBy(_ => RNG.Next())
+                                                        .Take(5)
+                                                        .ToList();
```
a property should not be returning different, *non-deterministic* results on subsequent accesses. even if it's test code. i would make this a method.

bdach commented:
the test scene is showing a one-frame flicker when the mods are starting to display.

https://user-images.githubusercontent.com/20418176/117539437-ec90b480-b00a-11eb-89f9-4b5a05aa423e.mp4

you can slow down to 0.1x or something to see it better.

bdach commented on osu.Game/Screens/Play/BeatmapMetadataDisplay.cs:
```diff
@@ -41,18 +43,22 @@ public bool Loading
         }
 
-        public BeatmapMetadataDisplay(WorkingBeatmap beatmap, Bindable<IReadOnlyList<Mod>> mods, Drawable facade)
+        public BeatmapMetadataDisplay(WorkingBeatmap beatmap, Bindable<IReadOnlyList<Mod>> mods, [CanBeNull] Drawable logoFacade)
```
is this ctor change for the test? it seems unnecessary, as you could pass in `Empty()` from the test instead of adding it in here.

bdach commented on osu.Game/Screens/Play/BeatmapMetadataDisplay.cs:
```diff
@@ -159,6 +178,13 @@ private void load()
+            starDifficulty = difficultyCache.GetBindableDifficulty(beatmap.BeatmapInfo);
+            starDifficulty.BindValueChanged(difficulty =>
+            {
+                if (difficulty.NewValue is StarDifficulty diff)
```
i don't personally like null-checks done with `is`. they look redundant. would just null-check normally.

frenzibyte replied:
well, I used it because I could shorten it from `difficulty.NewValue.Value` to `diff`, but I guess it's not a good pattern in general.

bdach commented on osu.Game/Screens/Play/BeatmapMetadataDisplay.cs:
```diff
@@ -66,11 +72,11 @@ private void load()
                             d.Anchor = Anchor.TopCentre;
                             d.Origin = Anchor.TopCentre;
-                        }),
+                        }) ?? Drawable.Empty(),
```
class name qualifier is redundant (but i'd like this to be reverted anyhow, see previous comment)

```suggestion
                        }) ?? Empty(),
```

bdach commented on osu.Game/Screens/Ranking/Expanded/StarRatingDisplay.cs:
```diff
@@ -78,32 +82,51 @@ private void load(OsuColour colours, BeatmapDifficultyCache difficultyCache)
+            string fractionPart = starRatingParts[1];
+            string separator = CultureInfo.CurrentCulture.NumberFormat.NumberDecimalSeparator;
+
+            background.Colour = difficulty.DifficultyRating == DifficultyRating.ExpertPlus
+                ? ColourInfo.GradientVertical(Color4Extensions.FromHex("#C1C1C1"), Color4Extensions.FromHex("#595959"))
+                : (ColourInfo)colours.ForDifficultyRating(difficulty.DifficultyRating);
```
not super sure about the lack of any sort of fade out transition here

frenzibyte replied:
to clarify, you mean fading colours of the background?

bdach replied:
that, and the text sprites

frenzibyte replied:
I've added proper transitioning along with rolling counter for the star rating display, I can imagine cases where the difficulty would need time to calculate before its shown, which would look bad on the player UX if it hides and shows back, moving the entire metadata flow back and forward. 

Not sure on the design of the placeholder though.

https://user-images.githubusercontent.com/22781491/117552234-c4c53f00-b052-11eb-9d42-2d7db1a5e5cf.mp4

peppy replied:
I'm not sure why this was added in this PR, but it's weird due to the font being right aligned and not fixed width.

peppy replied:
Please split this out into its own PR and fix that issue there. The bubble / text can't shift every frame of the transition.

frenzibyte replied:
Using fixed-width made it look really weird so I refrained from doing that, and I thought it wasn't that big of a change to split it, but can do.

peppy replied:
Well it's affecting a global component and you've updated it in a local usage. There's nothing to say you actually checked how this works in every other usage.

I'm not saying it should be fixed width, but if anything it should not be right aligned and should definitely not affect the size of the parent container to the point it's bouncing around.

frenzibyte replied:
I'm not really sure about how to fix this, I've checked fixed-width one more time to record a video of it and the results are indeed not appealing: (especially with `1.23` due to font being right-aligned as you mentioned)

https://user-images.githubusercontent.com/22781491/117606549-caaa4580-b162-11eb-8914-b3d5a930397c.mp4

Also fixing the width of the star rating display to 60px makes it look a bit questionable:

https://user-images.githubusercontent.com/22781491/117607349-756f3380-b164-11eb-898a-21e9ecb4067b.mp4

The `StarRatingDisplay` in general looks to be designed to not be changeable and make the numbers roll around, not sure how to proceed forward without leaving this as-is (or going with the second option of fixing width to 60px).

peppy replied:
Fixed width looks okay I think. If it's decided it doesn't then the numbers should not roll in the first place.

frenzibyte replied:
It shouldn't update all the time, only when the difficulty calculation is heavy, so I think it's fine to remove the rolling altogether and call it a day for this design.

frenzibyte committed:
Remove test scene description

frenzibyte committed:
Use regular test steps rather than one-time set up and scheduling

frenzibyte committed:
Remove nullable facade logic

frenzibyte committed:
Pass empty facade and replace random property with method instead

frenzibyte committed:
Remove pattern-matching on nullable with simple `.HasValue`/`.Value`

frenzibyte committed:
Apply further refactoring to star rating display UX-wise

frenzibyte committed:
Add initial fade in to the metadata display

Avoids first frame discrepancies from appearing in the test scene, those can be delt with later on, if needed.

frenzibyte committed:
Allow changing ruleset during test

frenzibyte commented:
> the test scene is showing a one-frame flicker when the mods are starting to display.
> 
>  2021-05-08.14-34-06.mp4 
> 
> you can slow down to 0.1x or something to see it better.

Resolved with initial fade in on test scenes, the underlying issues can be dealt with later on.

peppy commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapMetadataDisplay.cs:
```diff
@@ -0,0 +1,94 @@
+        }
+
+        [Test]
+        public void TestRandomFromDatabase()
```
really not sure about adding more test scenes which depend on the local beatmaps to function correctly

frenzibyte replied:
added it to check how metadata would look on real local beatmaps, and I imagined it to be useful for others (especially when it comes to tweaking the design).

unless you think it's better not to rely on local databases all the time and always go with the option of adding programmatically written beatmaps.

peppy committed:
Tidy up test scene

peppy committed:
Remove unnecessary ruleset switching steps

peppy committed:
Make test work without manually clicking things

peppy committed:
Add the osu! logo to the test scene

Makes no sense to add a test intended to test visual behaviour with one
of the main elements missing. Not sure how you would be able to test the
flow with the logo's presence.

peppy commented:
I've made some preliminary changes to the test scene to make it behave more like how it does within the game. Also added back the logo.

peppy committed:
Merge branch 'master' into player-loader-star-rating

peppy commented:
Make a test specifically for the star difficulty display, in the same PR where you split that functionality out.

frenzibyte commented:
Just to be extra clear, can you elaborate more on that? do you mean writing a new test scene that hooks up a star rating display to a bindable difficulty from the cache, on a written beatmap, to ensure it works?

bdach commented:
> simulates the case where star difficulty for a certain beatmap may take some time before the calculation finishes

That's the same thing as not setting difficulty initially and then setting it later after a delay, right? In that case, why not just skip interfacing with any beatmaps or difficulty caches whatsoever - or am I misunderstanding the intent?

peppy commented:
It's fine as is, but this PR still needs the removal of the animated counter.

frenzibyte committed:
Merge branch 'current-star-rating' into player-loader-star-rating

frenzibyte committed:
Make `BeatmapDifficultyCache.GetDifficultyAsync` virtual

frenzibyte committed:
Add test beatmap difficulty cache with calc. blocking support

frenzibyte committed:
Apply simple transforms to star rating display when ready if not

frenzibyte committed:
Merge branch 'current-star-rating' into player-loader-star-rating

frenzibyte committed:
Update usage due to nullability removal

peppy committed:
Merge branch 'master' into player-loader-star-rating

peppy reviewed (dismissed):
seems fine for now

frenzibyte committed:
Remove null conditional

Shouldn't guard against that here.
